{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GSD: Assessing ambiguous nts in PB 'complete' genomes\n",
    "\n",
    "Checks a collection of PacBio sequenced yeast 'complete' genomes from [Yue et al 2017](https://www.ncbi.nlm.nih.gov/pubmed/28416820) ambiguous/gap-representing residues. Comparing to the SGD reference sequence from [here](https://downloads.yeastgenome.org/sequence/S288C_reference/chromosomes/fasta/) as well. 'Complete' here is meant to refer to the analyses examing the mitochondrial and nuclear combined as one unit, unlike `GSD Assessing_ambiguous_nts_in_nuclear_PB_genomes.ipynb` where was just nuclear genome. And although Yue et al 2017 kept these two genomes separate, this is moving towards where will also analyze the 1011 genome collection where mitochondrial and nuclear genome mixed in one file. As will be seen here, the SGD reference sequence is easily available as one file with mitochondrial genome included; however, `GSD Assessing_ambiguous_nts_in_nuclear_PB_genomes.ipynb` shows it is fairly easy to put together one without the mitochondrial using the FASTA files of the individual chromosomes that are available from a different yeastgenome.org page.\n",
    "\n",
    "References for sequence data:  \n",
    "- [Contrasting evolutionary genome dynamics between domesticated and wild yeasts.\n",
    "Yue JX, Li J, Aigrain L, Hallin J, Persson K, Oliver K, BergstrÃ¶m A, Coupland P, Warringer J, Lagomarsino MC, Fischer G, Durbin R, Liti G. Nat Genet. 2017 Jun;49(6):913-924. doi: 10.1038/ng.3847. Epub 2017 Apr 17. PMID: 28416820](https://www.ncbi.nlm.nih.gov/pubmed/28416820)\n",
    "\n",
    "\n",
    "- [Life with 6000 genes. Goffeau A, Barrell BG, Bussey H, Davis RW, Dujon B, Feldmann H, Galibert F, Hoheisel JD, Jacq C, Johnston M, Louis EJ, Mewes HW, Murakami Y, Philippsen P, Tettelin H, Oliver SG. Science. 1996 Oct 25;274(5287):546, 563-7. PMID: 8849441](https://www.ncbi.nlm.nih.gov/pubmed/8849441)\n",
    "\n",
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation\n",
    "\n",
    "Get scripts and sequence data necessary.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pyfaidx\n",
      "  Downloading https://files.pythonhosted.org/packages/75/a5/7e2569527b3849ea28d79b4f70d7cf46a47d36459bc59e0efa4e10e8c8b2/pyfaidx-0.5.5.2.tar.gz\n",
      "Requirement already satisfied: six in /srv/conda/lib/python3.6/site-packages (from pyfaidx) (1.12.0)\n",
      "Requirement already satisfied: setuptools>=0.7 in /srv/conda/lib/python3.6/site-packages (from pyfaidx) (40.6.3)\n",
      "Building wheels for collected packages: pyfaidx\n",
      "  Running setup.py bdist_wheel for pyfaidx ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /home/jovyan/.cache/pip/wheels/54/a2/b4/e242e58d23b2808e191b214067880faa46cd2341f363886e0b\n",
      "Successfully built pyfaidx\n",
      "Installing collected packages: pyfaidx\n",
      "Successfully installed pyfaidx-0.5.5.2\n"
     ]
    }
   ],
   "source": [
    "!pip install pyfaidx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the genomes data by running these commands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3560      0 --:--:-- --:--:-- --:--:--  3560\n",
      "100 3687k  100 3687k    0     0  13.9M      0 --:--:-- --:--:-- --:--:-- 13.9M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   5235      0 --:--:-- --:--:-- --:--:--  5235\n",
      "100 22109  100 22109    0     0   146k      0 --:--:-- --:--:-- --:--:--  146k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   5085      0 --:--:-- --:--:-- --:--:--  5085\n",
      "100 3387k  100 3387k    0     0  14.1M      0 --:--:-- --:--:-- --:--:-- 35.5M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   5235      0 --:--:-- --:--:-- --:--:--  5393\n",
      "100 20363  100 20363    0     0   144k      0 --:--:-- --:--:-- --:--:--  144k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   5235      0 --:--:-- --:--:-- --:--:--  5393\n",
      "100 3348k  100 3348k    0     0  14.4M      0 --:--:-- --:--:-- --:--:-- 14.4M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3358      0 --:--:-- --:--:-- --:--:--  3358\n",
      "100 20829  100 20829    0     0   113k      0 --:--:-- --:--:-- --:--:--  113k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   2738      0 --:--:-- --:--:-- --:--:--  2781\n",
      "100 3406k  100 3406k    0     0  10.8M      0 --:--:-- --:--:-- --:--:-- 10.8M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   2870      0 --:--:-- --:--:-- --:--:--  2870\n",
      "100 21373  100 21373    0     0   107k      0 --:--:-- --:--:-- --:--:--  107k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   4944      0 --:--:-- --:--:-- --:--:--  4944\n",
      "100 3357k  100 3357k    0     0  12.1M      0 --:--:-- --:--:-- --:--:-- 12.1M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3560      0 --:--:-- --:--:-- --:--:--  3560\n",
      "100 21190  100 21190    0     0   133k      0 --:--:-- --:--:-- --:--:--  133k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   5741      0 --:--:-- --:--:-- --:--:--  5741\n",
      "100 3375k  100 3375k    0     0  12.2M      0 --:--:-- --:--:-- --:--:-- 12.2M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   2918      0 --:--:-- --:--:-- --:--:--  2966\n",
      "100 19885  100 19885    0     0   129k      0 --:--:-- --:--:-- --:--:--  129k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3122      0 --:--:-- --:--:-- --:--:--  3122\n",
      "100 3332k  100 3332k    0     0  13.5M      0 --:--:-- --:--:-- --:--:-- 13.5M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3490      0 --:--:-- --:--:-- --:--:--  3560\n",
      "100 18809  100 18809    0     0   129k      0 --:--:-- --:--:-- --:--:--  129k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3358      0 --:--:-- --:--:-- --:--:--  3296\n",
      "100 3658k  100 3658k    0     0  12.3M      0 --:--:-- --:--:-- --:--:-- 12.3M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   5235      0 --:--:-- --:--:-- --:--:--  5235\n",
      "100 18692  100 18692    0     0   146k      0 --:--:-- --:--:-- --:--:--  146k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   6846      0 --:--:-- --:--:-- --:--:--  7120\n",
      "100 3350k  100 3350k    0     0  15.0M      0 --:--:-- --:--:-- --:--:-- 52.4M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3560      0 --:--:-- --:--:-- --:--:--  3560\n",
      "100 18399  100 18399    0     0   126k      0 --:--:-- --:--:-- --:--:--  126k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3708      0 --:--:-- --:--:-- --:--:--  3708\n",
      "100 3349k  100 3349k    0     0  12.9M      0 --:--:-- --:--:-- --:--:-- 12.9M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   1854      0 --:--:-- --:--:-- --:--:--  1854\n",
      "100 18509  100 18509    0     0  80473      0 --:--:-- --:--:-- --:--:-- 80473\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   5235      0 --:--:-- --:--:-- --:--:--  5235\n",
      "100 3701k  100 3701k    0     0  15.1M      0 --:--:-- --:--:-- --:--:-- 15.1M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   7120      0 --:--:-- --:--:-- --:--:--  6846\n",
      "100 18886  100 18886    0     0   156k      0 --:--:-- --:--:-- --:--:--  156k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   3236      0 --:--:-- --:--:-- --:--:--  3236\n",
      "100 3362k  100 3362k    0     0  11.1M      0 --:--:-- --:--:-- --:--:-- 18.7M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   178  100   178    0     0   2119      0 --:--:-- --:--:-- --:--:--  2119\n",
      "100 17877  100 17877    0     0    98k      0 --:--:-- --:--:-- --:--:--  831k\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "# Prepare for getting PacBio (Yue et al 2017 sequences)\n",
    "#make a list of the strain designations\n",
    "yue_et_al_strains = [\"S288C\",\"DBVPG6044\",\"DBVPG6765\",\"SK1\",\"Y12\",\n",
    "                     \"YPS128\",\"UWOPS034614\",\"CBS432\",\"N44\",\"YPS138\",\n",
    "                     \"UFRJ50816\",\"UWOPS919171\"]\n",
    "# Get & unpack the genome sequences from strains \n",
    "for s in yue_et_al_strains:\n",
    "    !curl -LO http://yjx1217.github.io/Yeast_PacBio_2016/data/Nuclear_Genome/{s}.genome.fa.gz\n",
    "    !curl -OL http://yjx1217.github.io/Yeast_PacBio_2016/data/Mitochondrial_Genome/{s}.mt.genome.fa.gz\n",
    "    !gunzip -f {s}.genome.fa.gz\n",
    "    !gunzip -f {s}.mt.genome.fa.gz\n",
    "    # add the mitochondrial genome content onto the nuclear\n",
    "    !cat {s}.genome.fa {s}.mt.genome.fa > temp.genome.fa\n",
    "    !mv temp.genome.fa {s}.genome.fa\n",
    "!rm *.mt.genome.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "S288C.genome.fa chromosome identifiers tagged.\n",
      "DBVPG6044.genome.fa chromosome identifiers tagged.\n",
      "DBVPG6765.genome.fa chromosome identifiers tagged.\n",
      "SK1.genome.fa chromosome identifiers tagged.\n",
      "Y12.genome.fa chromosome identifiers tagged.\n",
      "YPS128.genome.fa chromosome identifiers tagged.\n",
      "UWOPS034614.genome.fa chromosome identifiers tagged.\n",
      "CBS432.genome.fa chromosome identifiers tagged.\n",
      "N44.genome.fa chromosome identifiers tagged.\n",
      "YPS138.genome.fa chromosome identifiers tagged.\n",
      "UFRJ50816.genome.fa chromosome identifiers tagged.\n",
      "UWOPS919171.genome.fa chromosome identifiers tagged."
     ]
    }
   ],
   "source": [
    "# add identifiers to each `chr` so results for each strain clear later\n",
    "chromosome_id_prefix = \"chr\"\n",
    "def add_strain_id_to_description_line(file,strain_id):\n",
    "    '''\n",
    "    Takes a file and edits every description line to add \n",
    "    strain_id after the caret.\n",
    "    \n",
    "    Saves the fixed file\n",
    "    '''\n",
    "    import sys\n",
    "    output_file_name = \"temp.txt\"\n",
    "    # prepare output file for saving so it will be open and ready\n",
    "    with open(output_file_name, 'w') as output_file:\n",
    "\n",
    "        # read in the input file\n",
    "        with open(file, 'r') as input_handler:\n",
    "            # prepare to give feeback later or allow skipping to certain start\n",
    "            lines_processed = 0\n",
    "\n",
    "            for line in input_handler:\n",
    "                lines_processed += 1\n",
    "                if line.startswith(\">\"):\n",
    "                    rest_o_line = line.split(\">\")\n",
    "                    new_line = \">\"+strain_id + rest_o_line[1]\n",
    "                else:\n",
    "                    new_line = line\n",
    "                \n",
    "                # Send text to output\n",
    "                output_file.write(new_line)\n",
    "\n",
    "    \n",
    "    # replace the original file with edited\n",
    "    !mv temp.txt {file}\n",
    "    # Feedback\n",
    "    sys.stderr.write(\"\\n{} chromosome identifiers tagged.\".format(file))\n",
    "\n",
    "for s in yue_et_al_strains:\n",
    "    add_strain_id_to_description_line(s+\".genome.fa\",s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 16.8M  100 16.8M    0     0  19.4M      0 --:--:-- --:--:-- --:--:-- 19.4M\n"
     ]
    }
   ],
   "source": [
    "# Get SGD reference sequence that includes nuclear and mitochondrial sequence as one file,\n",
    "# among others. I'll use file name for the reference genome worked out \n",
    "# in `GSD Assessing_ambiguous_nts_in_nuclear_PB_genomes.ipynb`, so more of the\n",
    "# previously worked out code will work.\n",
    "!curl -OL https://downloads.yeastgenome.org/sequence/S288C_reference/genome_releases/S288C_reference_genome_Current_Release.tgz\n",
    "!tar -xzf S288C_reference_genome_Current_Release.tgz\n",
    "!rm S288C_reference_genome_Current_Release.tgz\n",
    "!mv S288C_reference_genome_R64-2-1_20150113/S288C_reference_sequence_R64-2-1_20150113.fsa ./SGD_REF.genome.fa\n",
    "!rm -rf S288C_reference_genome_R64-2-1_20150113"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DBVPG6044.genome.fa',\n",
       " 'SGD_REF.genome.fa',\n",
       " 'DBVPG6765.genome.fa',\n",
       " 'CBS432.genome.fa',\n",
       " 'YPS128.genome.fa',\n",
       " 'UFRJ50816.genome.fa',\n",
       " 'S288C.genome.fa',\n",
       " 'SK1.genome.fa',\n",
       " 'UWOPS034614.genome.fa',\n",
       " 'UWOPS919171.genome.fa',\n",
       " 'YPS138.genome.fa',\n",
       " 'N44.genome.fa',\n",
       " 'Y12.genome.fa']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make a list of all `genome.fa` files, excluding `genome.fa.nhr` and `genome.fa.nin` and `genome.fansq`\n",
    "# The excluding was only necessary because I had run some BLAST queries preiminarily in development. Normally, \n",
    "# it would just be the `.re.fa` at the outset. (But keeping because removal ability could be useful.)\n",
    "fn_to_check = \"genome.fa\" \n",
    "genomes = []\n",
    "import os\n",
    "import fnmatch\n",
    "for file in os.listdir('.'):\n",
    "    if fnmatch.fnmatch(file, '*'+fn_to_check):\n",
    "        if not file.endswith(\".nhr\") and not file.endswith(\".nin\") and not file.endswith(\".nsq\") :\n",
    "            genomes.append(file)\n",
    "genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Now you are prepared to analyze each PacBio-sequenced genome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assessing the genomes in regards to stretches of N base calls\n",
    "\n",
    "### Total number of Ns present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are occurences of N in DBVPG6044.genome.fa.\n",
      "No ambiguous nucleotides or gaps in SGD_REF.genome.fa.\n",
      "There are occurences of N in DBVPG6765.genome.fa.\n",
      "There are occurences of N in CBS432.genome.fa.\n",
      "There are occurences of N in YPS128.genome.fa.\n",
      "There are occurences of N in UFRJ50816.genome.fa.\n",
      "There are occurences of N in S288C.genome.fa.\n",
      "There are occurences of N in SK1.genome.fa.\n",
      "There are occurences of N in UWOPS034614.genome.fa.\n",
      "There are occurences of N in UWOPS919171.genome.fa.\n",
      "There are occurences of N in YPS138.genome.fa.\n",
      "There are occurences of N in N44.genome.fa.\n",
      "There are occurences of N in Y12.genome.fa.\n"
     ]
    }
   ],
   "source": [
    "from pyfaidx import Fasta\n",
    "query = \"N\"\n",
    "#query = \"X\"\n",
    "#query = \"A\" # CONTROL. should occur!\n",
    "for g in genomes:\n",
    "    chrs = Fasta(g)\n",
    "    ambiguous = []\n",
    "    for x in chrs:\n",
    "        #print(x.name)\n",
    "        if query in str(x):\n",
    "            ambiguous.append(x.name)\n",
    "    #print(ambiguous)\n",
    "    if not ambiguous:\n",
    "        print (\"No ambiguous nucleotides or gaps in {}.\".format(g))\n",
    "    else:\n",
    "        print(\"There are occurences of {} in {}.\".format(query,g))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another way to assess, just count all the letters present:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 14.4 s, sys: 0 ns, total: 14.4 s\n",
      "Wall time: 15.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from pyfaidx import Fasta\n",
    "import pandas as pd\n",
    "import collections\n",
    "nt_counts = {}\n",
    "for g in genomes:\n",
    "    strain_id = g.split(\".genome.fa\")[0]\n",
    "    concatenated_seqs = \"\"\n",
    "    chrs = Fasta(g)\n",
    "    for x in chrs:\n",
    "        #print(x.name)\n",
    "        concatenated_seqs += str(x)\n",
    "    nt_counts[strain_id] = collections.Counter(concatenated_seqs)\n",
    "nt_count_df = pd.DataFrame.from_dict(nt_counts, orient='index').fillna(0)\n",
    "nt_count_df[\"Total_nts\"] = nt_count_df.sum(1)\n",
    "def percent_calc(items):\n",
    "    '''\n",
    "    takes a list of two items and calculates percentage of first item\n",
    "    within total (second item)\n",
    "    '''\n",
    "    return items[0]/items[1]\n",
    "nt_count_df['% N'] = nt_count_df[['N','Total_nts']].apply(percent_calc, axis=1)\n",
    "nt_count_df = nt_count_df.style.format({'Total_nts':'{:.2E}','% N':'{:.2%}'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "</style>  \n",
       "<table id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86\" > \n",
       "<thead>    <tr> \n",
       "        <th class=\"blank level0\" ></th> \n",
       "        <th class=\"col_heading level0 col0\" >A</th> \n",
       "        <th class=\"col_heading level0 col1\" >G</th> \n",
       "        <th class=\"col_heading level0 col2\" >T</th> \n",
       "        <th class=\"col_heading level0 col3\" >C</th> \n",
       "        <th class=\"col_heading level0 col4\" >N</th> \n",
       "        <th class=\"col_heading level0 col5\" >Total_nts</th> \n",
       "        <th class=\"col_heading level0 col6\" >% N</th> \n",
       "    </tr></thead> \n",
       "<tbody>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row0\" class=\"row_heading level0 row0\" >CBS432</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row0_col0\" class=\"data row0 col0\" >3715155</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row0_col1\" class=\"data row0 col1\" >2328075</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row0_col2\" class=\"data row0 col2\" >3705979</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row0_col3\" class=\"data row0 col3\" >2326244</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row0_col4\" class=\"data row0 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row0_col5\" class=\"data row0 col5\" >1.21E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row0_col6\" class=\"data row0 col6\" >0.14%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row1\" class=\"row_heading level0 row1\" >DBVPG6044</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row1_col0\" class=\"data row1 col0\" >3723086</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row1_col1\" class=\"data row1 col1\" >2293371</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row1_col2\" class=\"data row1 col2\" >3710750</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row1_col3\" class=\"data row1 col3\" >2294570</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row1_col4\" class=\"data row1 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row1_col5\" class=\"data row1 col5\" >1.20E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row1_col6\" class=\"data row1 col6\" >0.14%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row2\" class=\"row_heading level0 row2\" >DBVPG6765</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row2_col0\" class=\"data row2 col0\" >3681915</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row2_col1\" class=\"data row2 col1\" >2265102</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row2_col2\" class=\"data row2 col2\" >3666626</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row2_col3\" class=\"data row2 col3\" >2263910</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row2_col4\" class=\"data row2 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row2_col5\" class=\"data row2 col5\" >1.19E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row2_col6\" class=\"data row2 col6\" >0.15%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row3\" class=\"row_heading level0 row3\" >N44</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row3_col0\" class=\"data row3 col0\" >3650758</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row3_col1\" class=\"data row3 col1\" >2283907</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row3_col2\" class=\"data row3 col2\" >3646946</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row3_col3\" class=\"data row3 col3\" >2282669</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row3_col4\" class=\"data row3 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row3_col5\" class=\"data row3 col5\" >1.19E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row3_col6\" class=\"data row3 col6\" >0.15%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row4\" class=\"row_heading level0 row4\" >S288C</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row4_col0\" class=\"data row4 col0\" >3788833</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row4_col1\" class=\"data row4 col1\" >2331280</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row4_col2\" class=\"data row4 col2\" >3773075</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row4_col3\" class=\"data row4 col3\" >2332397</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row4_col4\" class=\"data row4 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row4_col5\" class=\"data row4 col5\" >1.22E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row4_col6\" class=\"data row4 col6\" >0.14%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row5\" class=\"row_heading level0 row5\" >SGD_REF</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row5_col0\" class=\"data row5 col0\" >3766349</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row5_col1\" class=\"data row5 col1\" >2317100</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row5_col2\" class=\"data row5 col2\" >3753080</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row5_col3\" class=\"data row5 col3\" >2320576</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row5_col4\" class=\"data row5 col4\" >0</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row5_col5\" class=\"data row5 col5\" >1.22E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row5_col6\" class=\"data row5 col6\" >0.00%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row6\" class=\"row_heading level0 row6\" >SK1</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row6_col0\" class=\"data row6 col0\" >3754961</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row6_col1\" class=\"data row6 col1\" >2315665</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row6_col2\" class=\"data row6 col2\" >3741582</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row6_col3\" class=\"data row6 col3\" >2318358</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row6_col4\" class=\"data row6 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row6_col5\" class=\"data row6 col5\" >1.21E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row6_col6\" class=\"data row6 col6\" >0.14%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row7\" class=\"row_heading level0 row7\" >UFRJ50816</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row7_col0\" class=\"data row7 col0\" >3769393</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row7_col1\" class=\"data row7 col1\" >2333235</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row7_col2\" class=\"data row7 col2\" >3763990</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row7_col3\" class=\"data row7 col3\" >2330694</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row7_col4\" class=\"data row7 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row7_col5\" class=\"data row7 col5\" >1.22E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row7_col6\" class=\"data row7 col6\" >0.14%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row8\" class=\"row_heading level0 row8\" >UWOPS034614</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row8_col0\" class=\"data row8 col0\" >3653129</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row8_col1\" class=\"data row8 col1\" >2250532</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row8_col2\" class=\"data row8 col2\" >3654099</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row8_col3\" class=\"data row8 col3\" >2246408</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row8_col4\" class=\"data row8 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row8_col5\" class=\"data row8 col5\" >1.18E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row8_col6\" class=\"data row8 col6\" >0.15%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row9\" class=\"row_heading level0 row9\" >UWOPS919171</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row9_col0\" class=\"data row9 col0\" >3684710</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row9_col1\" class=\"data row9 col1\" >2275952</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row9_col2\" class=\"data row9 col2\" >3677412</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row9_col3\" class=\"data row9 col3\" >2274358</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row9_col4\" class=\"data row9 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row9_col5\" class=\"data row9 col5\" >1.19E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row9_col6\" class=\"data row9 col6\" >0.15%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row10\" class=\"row_heading level0 row10\" >Y12</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row10_col0\" class=\"data row10 col0\" >3700059</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row10_col1\" class=\"data row10 col1\" >2278293</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row10_col2\" class=\"data row10 col2\" >3691795</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row10_col3\" class=\"data row10 col3\" >2276511</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row10_col4\" class=\"data row10 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row10_col5\" class=\"data row10 col5\" >1.20E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row10_col6\" class=\"data row10 col6\" >0.15%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row11\" class=\"row_heading level0 row11\" >YPS128</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row11_col0\" class=\"data row11 col0\" >3706321</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row11_col1\" class=\"data row11 col1\" >2285787</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row11_col2\" class=\"data row11 col2\" >3695971</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row11_col3\" class=\"data row11 col3\" >2282517</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row11_col4\" class=\"data row11 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row11_col5\" class=\"data row11 col5\" >1.20E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row11_col6\" class=\"data row11 col6\" >0.14%</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86level0_row12\" class=\"row_heading level0 row12\" >YPS138</th> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row12_col0\" class=\"data row12 col0\" >3660837</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row12_col1\" class=\"data row12 col1\" >2278232</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row12_col2\" class=\"data row12 col2\" >3653744</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row12_col3\" class=\"data row12 col3\" >2272948</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row12_col4\" class=\"data row12 col4\" >17357</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row12_col5\" class=\"data row12 col5\" >1.19E+07</td> \n",
       "        <td id=\"T_e4c466ea_1230_11e9_a27c_8e4158cefe86row12_col6\" class=\"data row12 col6\" >0.15%</td> \n",
       "    </tr></tbody> \n",
       "</table> "
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7f754ac3f4e0>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nt_count_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examining the number and size of stretches of Ns present"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a dataframe of the data and also make a summary one with the say, top five(?) number instances, for each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.16 s, sys: 0 ns, total: 3.16 s\n",
      "Wall time: 3.43 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# count frequency of blocks of Ns in the genomes\n",
    "import re\n",
    "from collections import defaultdict\n",
    "from pyfaidx import Fasta\n",
    "import pandas as pd\n",
    "import collections\n",
    "\n",
    "\n",
    "min_number_Ns_in_row_to_collect = 3\n",
    "pattern_obj = re.compile(\"N{{{},}}\".format(min_number_Ns_in_row_to_collect), re.I)  # adpated from\n",
    "# code worked out in `collapse_large_unknown_blocks_in_DNA_sequence.py`, which relied heavily on\n",
    "# https://stackoverflow.com/a/250306/8508004\n",
    "\n",
    "len_match_dict_by_strain = {}\n",
    "\n",
    "#genomes = [\"N44.genome.fa\"]\n",
    "\n",
    "for g in genomes:\n",
    "    len_match_dict = defaultdict(int)\n",
    "    strain_id = g.split(\".genome.fa\")[0]\n",
    "    records = Fasta(g)\n",
    "    for record in records:\n",
    "        for m in pattern_obj.finditer(str(record)):\n",
    "            len_match_dict[len(m.group())] += 1\n",
    "    #before passing final dictionary for strain to\n",
    "    #collection, add an entry of zero for size of 1 so the\n",
    "    #strain with no stretches will be in final dataframe\n",
    "    if not len_match_dict:\n",
    "        len_match_dict[1]=0\n",
    "    len_match_dict_by_strain[strain_id] = len_match_dict\n",
    "\n",
    "#stretches_size_freq_df = pd.DataFrame.from_dict(len_match_dict_by_strain, orient='index').fillna(0)\n",
    "#stretches_size_freq_df = pd.DataFrame.from_dict(len_match_dict_by_strain).fillna(0).stack().reset_index() # if wanted all stretches of size to show count even if none present\n",
    "stretches_size_freq_df = pd.DataFrame.from_dict(len_match_dict_by_strain).stack().reset_index() \n",
    "stretches_size_freq_df.columns = ['stretch_size','strain','#_instances'] \n",
    "stretches_size_freq_df = stretches_size_freq_df[['strain','stretch_size','#_instances']]\n",
    "stretches_size_freq_df = stretches_size_freq_df.sort_values(['stretch_size','#_instances'],ascending=[0,0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because of use of `%%time` in above cell, easiest to display resulting dataframe using next cell. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>stretch_size</th>\n",
       "      <th>#_instances</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DBVPG6044</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DBVPG6765</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CBS432</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YPS128</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UFRJ50816</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S288C</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SK1</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UWOPS034614</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>UWOPS919171</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>YPS138</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>N44</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Y12</td>\n",
       "      <td>17357</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SGD_REF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         strain  stretch_size  #_instances\n",
       "0     DBVPG6044         17357          1.0\n",
       "1     DBVPG6765         17357          1.0\n",
       "2        CBS432         17357          1.0\n",
       "3        YPS128         17357          1.0\n",
       "4     UFRJ50816         17357          1.0\n",
       "5         S288C         17357          1.0\n",
       "6           SK1         17357          1.0\n",
       "7   UWOPS034614         17357          1.0\n",
       "8   UWOPS919171         17357          1.0\n",
       "9        YPS138         17357          1.0\n",
       "10          N44         17357          1.0\n",
       "11          Y12         17357          1.0\n",
       "12      SGD_REF             1          0.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stretches_size_freq_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The totals overall of nucleotides are slight up; however, essentially same result as in `GSD Assessing_ambiguous_nts_in_nuclear_PB_genomes.ipynb` because no stretches of Ns (or any Ns) in the PacBio-sequenced mitochondria.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
